### PR TITLE
Improve sort performance using TournamentTree

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve sort performance using TournamentTree instead of BinaryHeap during merge of sorted data.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
The code LGTM
(need to add comments, remove commented out code, add specific perf tests if needed, adjust the remaining tests).

But order by decimals slowed down a lot: https://s3.amazonaws.com/clickhouse-test-reports/36078/7ca13bbabb5f701196b35ebdcc1c9bd46cb7e8d3/performance_comparison__actions__[1/4]/report.html

**Comment by kitaisreal** [comment]
For random integer data sort by 1 column:
```sql
SELECT rand64() as a FROM numbers(50000000) ORDER BY a FORMAT Null;
```
Before:
```sql
SELECT rand64() as a FROM numbers(50000000) ORDER BY a FORMAT Null;
0 rows in set. Elapsed: 12.322 sec. Processed 50.05 million rows, 400.37 MB (4.06 million rows/s., 32.49 MB/s.
```
After:
```sql
SELECT rand64() as a FROM numbers(50000000) ORDER BY a FORMAT Null;
0 rows in set. Elapsed: 8.665 sec. Processed 50.05 million rows, 400.37 MB (5.78 million rows/s., 46.20 MB/s.)
```

**Comment by kitaisreal** [comment]
For random big string table:
Table setup:
```sql
CREATE TABLE big_string_table (id UInt64, value String) ENGINE=MergeTree ORDER BY id;
INSERT INTO big_string_table SELECT number, repeat('t', 500) FROM system.numbers LIMIT 50000000;
```
Before:
```sql
SELECT id, value FROM big_string_table ORDER BY value FORMAT Null;
0 rows in set. Elapsed: 23.320 sec. Processed 50.00 million rows, 25.85 GB (2.14 million rows/s., 1.11 GB/s.)
```
After:
```sql
SELECT id, value FROM big_string_table ORDER BY value FORMAT Null;
0 rows in set. Elapsed: 14.461 sec. Processed 50.00 million rows, 25.85 GB (3.46 million rows/s., 1.79 GB/s.)
```

File: src/Common/TournamentTree.h [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        cursors[cursor_index].next();
+        replayGames(cursor_index);
+    }
+
+    ALWAYS_INLINE Cursor & getCursor(size_t cursor_index) { return cursors[cursor_index]; }
+
+    ALWAYS_INLINE void updateCursor(size_t cursor_index) { replayGames(cursor_index); }
+
+private:
+    ALWAYS_INLINE void buildTree()
```
Comment (antonio2368) [inline src/Common/TournamentTree.h:77]: This is pretty aggressive inlining.

Apart from affecting build time, did you measure the difference in performance with and without explicit inlining?
Too much inlining can negatively affect the performance so I'm just interested if you tested it out.

**Comment by alexey-milovidov** [comment]
One test has to be adapted to changed ordering.

**Comment by alexey-milovidov** [comment]
Also need to wait for #36361.

File: src/Common/examples/tournament_tree.cpp [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    {
+        std::cerr << "Wrong strategy expected 1 or 2" << std::endl;
+        return 1;
+    }
+
+    std::cout << "Compare count " << compare_count << std::endl;
+    compare_count = 0;
+
+    std::cout << "Sum " << sum_total << std::endl;
+    std::cout << "Elapsed " << watch.elapsedMicroseconds() << std::endl;
```
Comment (KochetovNicolai) [inline src/Common/examples/tournament_tree.cpp:280]: should we remove all `std::out`'s?

**Comment by kitaisreal** [comment]
@Mergifyio update

**Comment by alexey-milovidov** [comment]
Order by decimals is still slower than before.

**Comment by azat** [comment]
Any news on this one?

**Comment by alexey-milovidov** [comment]
@azat It is paused in favor of #31796

**Comment by MBkkt** [comment]
This PR gave me idea to improve same logic in our code.

I think it's pretty interesting and unexpecting that sift_down approach in average worse than pop_heap + push_heap

https://github.com/sweeneyde/multimerge/issues/5

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 0d840980ec83b46dedd9ab010dfe07175bf109f7 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/36078/0d840980ec83b46dedd9ab010dfe07175bf109f7/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by lgbo-ustc** [comment]
Is this will be merged ?

**Comment by kitaisreal** [comment]
> will this be merged ?

Currently `TournamentTree` implementation is slower than `Heap` in real world scenarios, when there are a lot of duplicate values when data is sorted. Currently we use `BatchHeap` implementation, check `SortCursor.h`.
It is possible to have performance improvements using `TournamentTree`, but we need to understand cardinality of sorted data, right now without statistics it is not possible.

**Comment by alexey-milovidov** [comment]
@kitaisreal, estimating cardinality is trivial - you take the first block and sample 10 random values from it. If they are different, you conclude that the cardinality is high.

**Comment by alexey-milovidov** [comment]
@kitaisreal, it's so sad that this is unfinished.
IMO, these performance optimizations should be merged.

**Pull request closed without merging**